collections/src/main/scala/strawman/collection/immutable/Map.scala

```diff
@@ -128,27 +128,37 @@ trait MapOps[K, +V, +CC[X, +Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]
-    def iterableFactory = underlying.iterableFactory
-    def mapFactory: MapFactory[Map] = underlying.mapFactory
-    protected[this] def mapFromIterable[K2, V2](it: collection.Iterable[(K2, V2)]): Map[K2,V2] = mapFactory.from(it)
-
-    // Specific building will keep the default but may lose the precise underlying type because our own V can be
-    // a supertype of the underlying collection's V so we cannot rebuild with potentially new values that are not
-    // valid for the underlying collection.
-    protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): Map[K,V] = new WithDefault[K, V](mapFactory.from(coll), d)
-    protected[this] def newSpecificBuilder(): mutable.Builder[(K, V), Map[K,V]] =
-      mapFactory.newBuilder[K, V]().mapResult(new WithDefault[K, V](_, d))
-    override def size = underlying.size
-    def get(key: K) = underlying.get(key)
-    def iterator() = underlying.iterator()
-    override def default(key: K): V = d(key)
-    override def empty = new WithDefault(underlying.empty, d)
-    override def updated[V1 >: V](key: K, value: V1): WithDefault[K, V1] = new WithDefault[K, V1](underlying.updated[V1](key, value), d)
-    override def remove (key: K): WithDefault[K, V] = new WithDefault(underlying - key, d)
-    override def withDefault[V1 >: V](d: K => V1): immutable.Map[K, V1] = new WithDefault[K, V1](underlying, d)
-    override def withDefaultValue[V1 >: V](d: V1): immutable.Map[K, V1] = new WithDefault[K, V1](underlying, x => d)
+  object WithDefault extends MapFactory.Delegate[Map](HashMap)
+
+  final class WithDefault[K, +V](val underlying: Map[K, V], val defaultValue: K => V)
+    extends Map[K, V]
+      with WithDefaultOps[K, V, Map[K, V]] {
+
+    override def default(key: K): V = defaultValue(key)
+
+    def remove(key: K): Map.WithDefault[K, V] = new WithDefault[K, V](underlying - key, defaultValue)
+
+    def updated[V1 >: V](key: K, value: V1): Map.WithDefault[K, V1] =
+      new WithDefault[K, V1](underlying.updated[V1](key, value), defaultValue)
+
+    def empty: Map.WithDefault[K, V] = new WithDefault[K, V](underlying.empty, defaultValue)
+
+    protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): Map.WithDefault[K, V] =
+      new WithDefault[K, V](mapFactory.from(coll), defaultValue)
+
+    protected[this] def newSpecificBuilder(): Builder[(K, V), Map.WithDefault[K, V]] =
+      mapFactory.newBuilder[K, V]().mapResult(new WithDefault[K, V](_, defaultValue))
+
+    override def mapFactory: MapFactory[Map] = WithDefault
+
+    override def get(key: K): Option[V] = underlying.get(key)
+
+    override def iterableFactory: IterableFactoryLike[Iterable] = underlying.iterableFactory
+
+    override def iterator(): strawman.collection.Iterator[(K, V)] = underlying.iterator()
+
+    override protected[this] def mapFromIterable[K2, V2](it: strawman.collection.Iterable[(K2, V2)]): Map[K2, V2] =
+      Map.from[K2, V2](it)
```

Comment:
We could use `underlying.mapFactory.from`, to preserve the same map implementation.